dev-docs/RFCs/v6.0/offscreen-render-rfc.md

```diff
@@ -0,0 +1,100 @@
+
+
+## State of the API
+
+OffscreenCanvas is in active development (behind experimental flags) in both FF and [Chrome](https://www.chromestatus.com/feature/5424182347169792). In Chrome, go to `chrome://flags` and enable "Experimental web platform features."
+
+The API still seems limited/unstable:
+- FF only support OffscreenCanvas with WebGL1 context.
+- Chrome's render buffer does not update in some test scenarios.
+
+
+## Proposal: OffscreenAnimationLoop
+
+Add an `OffscreenAnimationLoop` that mirrors the `AnimationLoop` class but delegate rendering to a worker.
+
+Static methods:
+- `createWorker(opts)` - returns script that can be used to spin up a new Worker. `opts` are the options used to construct an `AnimationLoop` instance on the worker thread.
+
+Constructor:
+- `new OffscreenAnimationLoop(opts)` - `opts` are the same options used to construct an `AnimationLoop`, with additional fields:
+    + `opts.worker` (Worker) - the worker instance running a script created by `OffscreenAnimationLoop.createWorker`
+
+Methods:
+- `setProps(props)` - equivalent to `animationLoop.setProps(props)`
+- `start(opts)` - equivalent to `animationLoop.start(opts)`
+- `stop()` - equivalent to `animationLoop.stop()`
+
+
+### Examples
+
+```
+/// animation.js
+import {GL, Model, Framebuffer, Matrix4} from 'luma.gl';
+
+export default {
+    onInitialize: () => {
+        // set up scene
+    },
+    onRender: () => {
+        // render frame
+    },
+    glOptions: {},
+    autoResizeDrawingBuffer: true
+};
+```
+
+Render on main thread:
+```
+/// app.js
+import animationLoopOptions from './animation';
+import {AnimationLoop} from 'luma.gl';
+
+const animationLoop = new AnimationLoop(animationLoopOptions);
+animationLoop.start({canvas: 'lumagl-canvas'});
+```
+
+Render on worker thread:
+```
+/// app.js
+import animationLoopOptions from './animation';
+import {OffscreenAnimationLoop} from 'luma.gl';
+import webworkify from 'webworkify';
+
+const worker = OffscreenAnimationLoop.createWorker(animationLoopOptions);
+const animationLoop = new OffscreenAnimationLoop({
+    ...animationLoopOptions,
+    worker: webworkify(worker)
+});
+animationLoop.start({canvas: 'lumagl-canvas'});
+```
+
+
+## Proposal: Interactivity
+
+`AnimationLoop` does not directly handle user input (mousemove, click, etc.). This is usually not an issue as the user can update their own application state from event callbacks and then access that state from within `onRender`. In the offscreen rendering case, this is no longer possible as `onRender` is on a different thread from where the event callbacks are executed.
+
+The proposed solution is to expose a new API on both `AnimationLoop` and `OffscreenAnimationLoop` for applications to send custom data to the renderer:
+
+- Add `setUserData` method to `AnimationLoop`. It adds a field `userData` to `animationLoop._callbackData` which can be accessed in the `onRender` callback.
```

Comment:
Maybe an interface for forwarding some kind of serialized `mjolnir.js` events?

Reply:
That sounds arbitrary as `mjolnir.js` is not a luma.gl dependency.